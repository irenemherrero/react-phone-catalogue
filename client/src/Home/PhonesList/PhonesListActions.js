import { fetchPhones } from '../../services/phonesService'

export const ERROR_FETCHING = 'ERROR_FETCHING'
export const REQUEST_PHONES = 'REQUEST_PHONES'
export const RECEIVE_PHONES = 'RECEIVE_PHONES'

export const requestPhones = () => ({
  type: REQUEST_PHONES,
})

export const receivePhones = (data) => ({
  type: RECEIVE_PHONES,
  phones: data,
})

export const errorFetching = (error) => ({
  type: ERROR_FETCHING,
  error: error,
})

export const getPhones = (dispatch) => {
  dispatch(requestPhones())
  return fetchPhones()
    .then((res) => dispatch(receivePhones(res.data)))
    .catch((err) => dispatch(errorFetching(err.message)))
}
