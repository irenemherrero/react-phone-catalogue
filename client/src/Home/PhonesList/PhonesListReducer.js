import {
  REQUEST_PHONES,
  RECEIVE_PHONES,
  ERROR_FETCHING,
} from './PhonesListActions'

const initialState = {
  isFetching: false,
  phones: [],
  error: null,
}

const phonesReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_PHONES:
      return {
        ...state,
        isFetching: true,
      }
    case RECEIVE_PHONES:
      return {
        ...state,
        isFetching: false,
        phones: action.phones,
      }
    case ERROR_FETCHING:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    default:
      return state
  }
}

export { phonesReducer }
