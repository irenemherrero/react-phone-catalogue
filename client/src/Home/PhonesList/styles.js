import styled from 'styled-components'
import tw from 'tailwind.macro'

const ResponsiveContainer = styled.div`
  ${tw`w-full md:w-1/2 lg:w-1/3 p-2`}
`
const PhoneCard = styled.div`
  ${tw`flex flex-row shadow-md bg-white cursor-pointer
      h-48 max-w-xs mx-auto 
      hover:shadow-xl`}
  &:hover {
    transform: scale(1.05);
  }
`

const PhoneCardInfo = styled.section`
  ${tw`flex flex-col pt-6 w-full`}
  h1 {
    ${tw`font-semibold text-xl`}
  }
  h2 {
    ${tw`text-gray-600`}
  }
  footer {
    ${tw`self-end m-6 font-semibold text-secondary text-lg`}
  }
`

export { ResponsiveContainer, PhoneCard, PhoneCardInfo }
