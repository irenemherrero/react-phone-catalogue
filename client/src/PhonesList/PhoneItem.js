import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import tw from 'tailwind.macro'

const ResponsiveContainer = styled.div`
  ${tw`w-full md:w-1/2 lg:w-1/3 px-2 my-2`}
`
const PhoneCard = styled.div`
  ${tw`flex flex-row shadow-md bg-white cursor-pointer hover:shadow-xl`}
  &:hover {
    transform: scale(1.04);
  }
`
const PhoneCardImage = styled.img`
  ${tw`h-48 w-40`}
`
const PhoneCardInfo = styled.div`
  ${tw`flex flex-col p-6`}
  h1 {
    ${tw`font-semibold text-lg`}
  }
  p {
    ${tw`text-gray-600`}
  }
`
const ColorsContainer = styled.div`
  ${tw`flex flex-row mt-6`}
`

const PhoneColor = styled.span`
  ${tw`w-10 h-5 mr-1 border rounded`}
  background-color: ${(props) => props.color}
`

const PhoneItem = (props) => {
  const { id, name, onClick, imageFileName, price, color } = props
  const urlImage = process.env.PHONES_IMAGES_URL + imageFileName
  return (
    <ResponsiveContainer onClick={() => onClick(id)}>
      <PhoneCard>
        <div className="flex-shrink-0">
          <PhoneCardImage src={urlImage} alt={name} />
        </div>
        <PhoneCardInfo>
          <h1>{name}</h1>
          <p>{price}â‚¬</p>
          <ColorsContainer>
            {color.split(',').map((c) => (
              <PhoneColor key={c} color={c} />
            ))}
          </ColorsContainer>
        </PhoneCardInfo>
      </PhoneCard>
    </ResponsiveContainer>
  )
}

PhoneItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  imageFileName: PropTypes.string,
  price: PropTypes.number,
  color: PropTypes.string,
}
PhoneItem.defaultProps = {
  imageFileName: 'noimage.png',
  price: 0,
  onClick: () => {},
}

export { PhoneItem }
