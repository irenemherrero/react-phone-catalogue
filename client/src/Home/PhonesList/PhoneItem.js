import React from 'react'
import PropTypes from 'prop-types'
import { AvailableColors, PhoneImage } from '../../common'
import { ResponsiveWrapper, PhoneCard, PhoneCardInfo } from './styles'

/**
 * Showing phone item
 * @component
 */
const PhoneItem = (props) => {
  const { id, name, manufacturer, color, price, imageFileName, onClick } = props
  const urlImage = process.env.PHONES_IMAGES_URL + imageFileName
  return (
    <ResponsiveWrapper onClick={() => onClick(id)}>
      <PhoneCard>
        <PhoneImage src={urlImage} alt={name} />
        <PhoneCardInfo>
          <h1>{name}</h1>
          <h2>{manufacturer}</h2>
          <AvailableColors colors={color.split(',')} />
          <footer>{`${price} â‚¬`}</footer>
        </PhoneCardInfo>
      </PhoneCard>
    </ResponsiveWrapper>
  )
}

PhoneItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  manufacturer: PropTypes.string,
  /**comma separated colors */
  color: PropTypes.string,
  price: PropTypes.number,
  imageFileName: PropTypes.string,
  /**fires parent onClick */
  onClick: PropTypes.func,
}
PhoneItem.defaultProps = {
  manufacturer: 'N/A',
  imageFileName: 'noimage.png',
  price: 0,
  onClick: () => {},
}

export { PhoneItem }
