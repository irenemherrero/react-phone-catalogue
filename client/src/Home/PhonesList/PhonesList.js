import React from 'react'
import { useHistory } from 'react-router-dom'
import PropTypes from 'prop-types'
import { PhoneItem } from './PhoneItem'
import styled from 'styled-components'
import tw from 'tailwind.macro'

const ItemsWrapper = styled.div`
  ${tw`flex flex-wrap justify-center       
       py-2 px-0 sm:px-8 
       max-w-6xl`}
`
/**
 * Component for showing list of phones
 * @component
 */
const PhonesList = (props) => {
  const { phones } = props
  const history = useHistory()
  const showDetail = (id) => {
    history.push(`/phone/${id}`)
  }
  return (
    <ItemsWrapper>
      {phones.map((phone) => (
        <PhoneItem {...phone} key={phone.id} onClick={showDetail} />
      ))}
    </ItemsWrapper>
  )
}

PhonesList.propTypes = {
  /**
   * list of phones
   */
  phones: PropTypes.array.isRequired,
}

export { PhonesList }
