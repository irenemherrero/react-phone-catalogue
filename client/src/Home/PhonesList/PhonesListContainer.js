import React, { useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { getPhones } from './PhonesListActions'
import { PhoneList } from './PhonesList'
import { CenterContentWrapper } from '@/common/CenterContentWrapper'
import { Loader } from '@/common/Loader'

const PhonesListContainer = () => {
  const { isFetching, phones, error } = useSelector((state) => state)
  const dispatch = useDispatch()
  useEffect(() => {
    getPhones(dispatch)
  }, [])

  // return <Loader />
  return (
    <CenterContentWrapper>
      {error ? (
        <span>{error}</span>
      ) : isFetching ? (
        <Loader />
      ) : (
        <PhoneList phones={phones} />
      )}
    </CenterContentWrapper>
  )
}

export { PhonesListContainer }
