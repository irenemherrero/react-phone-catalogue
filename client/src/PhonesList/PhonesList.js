import React from 'react'
import { useHistory } from 'react-router-dom'
import PropTypes from 'prop-types'
import { PhoneItem } from './PhoneItem'
import styled from 'styled-components'
import tw from 'tailwind.macro'

const CardsContainer = styled.div`
  ${tw`flex flex-wrap my-4 mx-16 md:mx-8 lg:mx-1 xl:mx-16`}
`

const PhoneList = (props) => {
  const { phones } = props
  const history = useHistory()
  const showDetail = (id) => {
    history.push(`/phone/${id}`)
  }
  return (
    <CardsContainer>
      {phones.map((phone) => (
        <PhoneItem {...phone} key={phone.id} onClick={showDetail} />
      ))}
    </CardsContainer>
  )
}

PhoneList.propTypes = {
  phones: PropTypes.array.isRequired,
}

export { PhoneList }
