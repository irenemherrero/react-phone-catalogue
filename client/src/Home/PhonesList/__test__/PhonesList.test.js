import React from 'react'
import { PhonesList } from '../PhonesList'
import { render, cleanup } from '@testing-library/react'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}))

const mockPhones = [
  {
    id: 0,
    name: 'iPhone 7',
    manufacturer: 'Apple',
    description:
      'iPhone 7 dramatically improves the most important aspects of the iPhone experience. It introduces advanced new camera systems. The best performance and battery life ever in an iPhone. Immersive stereo speakers. The brightest, most colorful iPhone display. Splash and water resistance*. And it looks every bit as powerful as it is. This is iPhone 7.',
    color: 'black,blue',
    price: 769,
    imageFileName: 'IPhone_7.png',
    screen: '4,7 inch IPS',
    processor: 'A10 Fusion',
    ram: 2,
  },
  {
    id: 1,
    name: 'Galaxy S7',
    manufacturer: 'Samsung',
    description:
      'Introducing the smartphone your life can not do without, The Samsung Galaxy S7. Packed with features to keep you both productive and entertained, all in a sleek, slim design that fits comfortably in your hand or pocket.',
    color: 'black,blue,grey',
    price: 209,
    imageFileName: 'Galaxy_S7.png',
    screen: '5,1 inch Quad-HD',
    processor: 'Snapdragon 820',
    ram: 4,
  },
  {
    id: 2,
    name: 'Honor 10',
    manufacturer: 'Huawei',
    description:
      'Great phone with an excellent interface. One of the best mid price range phones in the market.',
    color: 'blue,white',
    price: 399,
    imageFileName: 'Honor_10.png',
    screen: '5,84 inch Full-HD',
    processor: 'Kirin 970',
    ram: 6,
  },
]

describe('PhonesList (a list of PhoneItems)', () => {
  afterEach(cleanup)

  it('render correctly', () => {
    const component = render(<PhonesList phones={mockPhones} />)
    expect(component).toMatchSnapshot()
  })
  it('num of items', () => {
    const { container } = render(<PhonesList phones={mockPhones} />)
    expect(container.firstChild.children.length).toEqual(mockPhones.length)
  })
})
