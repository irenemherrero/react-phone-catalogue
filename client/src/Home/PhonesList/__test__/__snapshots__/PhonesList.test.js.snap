// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhonesList (a list of PhoneItems) render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-fzplWN dYqHOa"
      >
        <div
          class="sc-fzozJi SWbVy"
        >
          <div
            class="sc-fzoLsD bjHBzu"
          >
            <img
              alt="iPhone 7"
              class="sc-Axmtr cFtcXX object-cover"
              src="undefinedIPhone_7.png"
            />
            <section
              class="sc-fzpans jtdzVu"
            >
              <h1>
                iPhone 7
              </h1>
              <h2>
                Apple
              </h2>
              <div
                class="sc-AxhCb JKePI"
              >
                <span
                  class="sc-AxhUy eyJlmj"
                  color="black"
                />
                <span
                  class="sc-AxhUy deRsvS"
                  color="blue"
                />
              </div>
              <footer>
                769 €
              </footer>
            </section>
          </div>
        </div>
        <div
          class="sc-fzozJi SWbVy"
        >
          <div
            class="sc-fzoLsD bjHBzu"
          >
            <img
              alt="Galaxy S7"
              class="sc-Axmtr cFtcXX object-cover"
              src="undefinedGalaxy_S7.png"
            />
            <section
              class="sc-fzpans jtdzVu"
            >
              <h1>
                Galaxy S7
              </h1>
              <h2>
                Samsung
              </h2>
              <div
                class="sc-AxhCb JKePI"
              >
                <span
                  class="sc-AxhUy eyJlmj"
                  color="black"
                />
                <span
                  class="sc-AxhUy deRsvS"
                  color="blue"
                />
                <span
                  class="sc-AxhUy bdkWcF"
                  color="grey"
                />
              </div>
              <footer>
                209 €
              </footer>
            </section>
          </div>
        </div>
        <div
          class="sc-fzozJi SWbVy"
        >
          <div
            class="sc-fzoLsD bjHBzu"
          >
            <img
              alt="Honor 10"
              class="sc-Axmtr cFtcXX object-cover"
              src="undefinedHonor_10.png"
            />
            <section
              class="sc-fzpans jtdzVu"
            >
              <h1>
                Honor 10
              </h1>
              <h2>
                Huawei
              </h2>
              <div
                class="sc-AxhCb JKePI"
              >
                <span
                  class="sc-AxhUy deRsvS"
                  color="blue"
                />
                <span
                  class="sc-AxhUy beRWZT"
                  color="white"
                />
              </div>
              <footer>
                399 €
              </footer>
            </section>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-fzplWN dYqHOa"
    >
      <div
        class="sc-fzozJi SWbVy"
      >
        <div
          class="sc-fzoLsD bjHBzu"
        >
          <img
            alt="iPhone 7"
            class="sc-Axmtr cFtcXX object-cover"
            src="undefinedIPhone_7.png"
          />
          <section
            class="sc-fzpans jtdzVu"
          >
            <h1>
              iPhone 7
            </h1>
            <h2>
              Apple
            </h2>
            <div
              class="sc-AxhCb JKePI"
            >
              <span
                class="sc-AxhUy eyJlmj"
                color="black"
              />
              <span
                class="sc-AxhUy deRsvS"
                color="blue"
              />
            </div>
            <footer>
              769 €
            </footer>
          </section>
        </div>
      </div>
      <div
        class="sc-fzozJi SWbVy"
      >
        <div
          class="sc-fzoLsD bjHBzu"
        >
          <img
            alt="Galaxy S7"
            class="sc-Axmtr cFtcXX object-cover"
            src="undefinedGalaxy_S7.png"
          />
          <section
            class="sc-fzpans jtdzVu"
          >
            <h1>
              Galaxy S7
            </h1>
            <h2>
              Samsung
            </h2>
            <div
              class="sc-AxhCb JKePI"
            >
              <span
                class="sc-AxhUy eyJlmj"
                color="black"
              />
              <span
                class="sc-AxhUy deRsvS"
                color="blue"
              />
              <span
                class="sc-AxhUy bdkWcF"
                color="grey"
              />
            </div>
            <footer>
              209 €
            </footer>
          </section>
        </div>
      </div>
      <div
        class="sc-fzozJi SWbVy"
      >
        <div
          class="sc-fzoLsD bjHBzu"
        >
          <img
            alt="Honor 10"
            class="sc-Axmtr cFtcXX object-cover"
            src="undefinedHonor_10.png"
          />
          <section
            class="sc-fzpans jtdzVu"
          >
            <h1>
              Honor 10
            </h1>
            <h2>
              Huawei
            </h2>
            <div
              class="sc-AxhCb JKePI"
            >
              <span
                class="sc-AxhUy deRsvS"
                color="blue"
              />
              <span
                class="sc-AxhUy beRWZT"
                color="white"
              />
            </div>
            <footer>
              399 €
            </footer>
          </section>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
